android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandlerMotionEventAdapter.java

```diff
@@ -0,0 +1,177 @@
+    mHandler = handler;
+    mActivePointers = new HashSet<>();
+  }
+
+  private GestureHandlerMotionEventAdapter(GestureHandlerMotionEventAdapter other) {
+    mEvent = MotionEvent.obtain(other.mEvent);
+    mActivePointers = other.mActivePointers;
+    mHandler = other.mHandler;
+  }
+
+  static GestureHandlerMotionEventAdapter obtain(GestureHandlerMotionEventAdapter other) {
+    return new GestureHandlerMotionEventAdapter(other);
+  }
+
+  public int getActionMasked() {
+    int action = mEvent.getActionMasked();
+
+    if (mActivePointers.size() == 0)
+      Log.e("GHMEAdapter", "Asked for empty event");
+
+    if (action == MotionEvent.ACTION_POINTER_DOWN && mActivePointers.size() == 1) {
+      // handle when many fingers on screen but only one just touched the area
+      return MotionEvent.ACTION_DOWN;
+    }
+
+    if (action == MotionEvent.ACTION_POINTER_UP && (mActivePointers.size() == 1)) {
+      // handle when many fingers on screen but the one of active was removed from area
+      return MotionEvent.ACTION_UP;
+    }
+
+    return mEvent.getActionMasked();
+  }
+
+  public int getPointerCount() {
+    return mActivePointers.size();
+  }
+
+  public int getMotionEventPointerCount() {
+    return mEvent.getPointerCount();
+  }
+
+  public boolean containsIndexOfMotionEvent(int index) {
+    return mActivePointers.contains(mEvent.getPointerId(index));
+  }
+
+  public float getX() {
+    return mEvent.getX();
+  }
+
+  public float getY() {
+    return mEvent.getY();
+  }
+
+  public float getRawX() {
+    return mEvent.getRawX();
+  }
+
+  public float getRawY() {
+    return mEvent.getRawY();
+  }
+
+  public int getActionIndex() {
+    return mEvent.getActionIndex();
+  }
+
+  public int getFirstPoinerId() {
+    return mFirstPointerId;
+  }
+
+  public float getY(int pointerIndex) {
+    return mEvent.getY(pointerIndex);
+  }
+
+  public float getX(int pointerIndex) {
+    return mEvent.getX(pointerIndex);
+  }
+
+  public MotionEvent getRawEvent() {
+    return mEvent;
+  }
+
+  public void setVelocityTracker(VelocityTracker velocityTracker) {
+    mVelocityTracker = velocityTracker;
+  }
+
+  public float getXVelocity() {
+    float sum = 0;
+    for (int i : mActivePointers) {
+      sum += mVelocityTracker.getXVelocity(i);
+    }
+    return sum / mActivePointers.size();
+  }
+
+  public float getYVelocity() {
+    float sum = 0;
+    for (int i : mActivePointers) {
+      sum += mVelocityTracker.getYVelocity(i);
+    }
+    return sum / mActivePointers.size();
+  }
+
+  public void reset() {
+    mActivePointers.clear();
+    mFirstPointerId = -1;
+  }
+
+
+  public boolean updateMotionEventBeforeHandling(MotionEvent event) {
+    int action = event.getActionMasked();
+    if (action == MotionEvent.ACTION_POINTER_DOWN || action == MotionEvent.ACTION_DOWN) {
+      int index = event.getActionIndex();
+      if (mHandler.isWithinBounds(mHandler.getView(), event.getX(index), event.getY(index))) {
```

Comment:
move this logic to orchestrator `deliverEventToGestureHandler` such that we only call `handler.handle(event)` if event is of an interest to the given handler.

In this case we no longer need to have handler reference in here